flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -417,12 +417,55 @@ public void cancel() {
 	@Override
 	public final boolean updateTaskExecutionState(final TaskExecutionState taskExecutionState) {
 		final Optional<ExecutionVertexID> executionVertexId = getExecutionVertexId(taskExecutionState.getID());
-		if (executionVertexId.isPresent()) {
-			executionGraph.updateState(taskExecutionState);
-			updateTaskExecutionStateInternal(executionVertexId.get(), taskExecutionState);
+
+		boolean updateSuccess = executionGraph.updateState(taskExecutionState);
+
+		if (updateSuccess && executionVertexId.isPresent()) {
+			Optional<TaskExecutionState> validTaskExecutionState = getValidTaskExecutionState(
+				executionVertexId.get(),
+				taskExecutionState);
+
+			if (validTaskExecutionState.isPresent()) {
+				updateTaskExecutionStateInternal(executionVertexId.get(), validTaskExecutionState.get());
+			}
 			return true;
+		} else {
+			return false;
+		}
+	}
+
+	private Optional<TaskExecutionState> getValidTaskExecutionState(
+			final ExecutionVertexID executionVertexId,
+			final TaskExecutionState taskExecutionState) {
+
+		final ExecutionVertex executionVertex = getExecutionVertex(executionVertexId);
+		final ExecutionAttemptID currentExecutionAttemptID = executionVertex.getCurrentExecutionAttempt().getAttemptId();
+
+		// check whether this state update is outdated
+		if (!currentExecutionAttemptID.equals(taskExecutionState.getID())) {
+			return Optional.empty();
 		}
-		return false;
+
+		// only notifies FINISHED and FAILED states which are needed at the moment.
+		// can be refined in FLINK-14233 after the legacy scheduler is removed and
+		// the actions are factored out from ExecutionGraph.
+		switch (taskExecutionState.getExecutionState()) {
```

Comment:
Would there be harm to notify all states?

Reply:
I think it would be simpler to only consider these two required state changes.
Notifying other states might be no hard but I have not thought it thoroughly. Moreover, it's hard to notify all states cleanly at the moment before we factored all actions out from ExecutionGraph. So I'd prefer to do it in a follow up ticket in Flink 1.11, i.e. FLINK-14233

Reply:
Ok